connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java

```diff
@@ -53,36 +56,68 @@
 
     private Map<String, String> config;
     private State state;
+    private final OffsetStorageReader offsetStorageReader;
 
-    public WorkerConnector(String connName,
-                           Connector connector,
-                           ConnectorContext ctx,
-                           ConnectorStatus.Listener statusListener) {
+    public WorkerConnector(final String connName,
+                           final Connector connector,
+                           final ConnectorContext ctx,
+                           final ConnectorStatus.Listener statusListener,
+                           final OffsetStorageReader offsetStorageReader) {
         this.connName = connName;
         this.ctx = ctx;
         this.connector = connector;
         this.statusListener = statusListener;
         this.state = State.INIT;
+        this.offsetStorageReader = offsetStorageReader;
     }
 
     public void initialize(ConnectorConfig connectorConfig) {
         try {
             this.config = connectorConfig.originalsStrings();
             log.debug("{} Initializing connector {} with config {}", this, connName, config);
-
-            connector.initialize(new ConnectorContext() {
+            final ConnectorContext delegateCtx = new ConnectorContext() {
                 @Override
                 public void requestTaskReconfiguration() {
-                    ctx.requestTaskReconfiguration();
+                    WorkerConnector.this.ctx.requestTaskReconfiguration();
                 }
 
                 @Override
                 public void raiseError(Exception e) {
                     log.error("{} Connector raised an error", this, e);
                     onFailure(e);
-                    ctx.raiseError(e);
+                    WorkerConnector.this.ctx.raiseError(e);
                 }
-            });
+            };
+            if (isSinkConnector()) {
+                connector.initialize(new SinkConnectorContext() {
+                    @Override
+                    public void requestTaskReconfiguration() {
+                        delegateCtx.requestTaskReconfiguration();
```

Comment:
Why use the delegate? Why not just call the methods on the `WorkerConnector`'s fields from within the `SourceConnectorContext` and `SinkConnectorContext` methods? E.g.,

      @Override
      public void requestTaskReconfiguration() {
           ctx.requestTaskReconfiguration();
      }

Reply:
It is to keep the same logic as the raiseError() method. The delegate implements the ConnectorContext.

Reply:
ok

Reply:
How about we consider using three nested classes to make this logic a bit easier to read?